protocols/request-response/src/lib.rs

```diff
@@ -431,9 +431,14 @@ where
     /// [`PeerId`] initiated by [`RequestResponse::send_request`] is still
     /// pending, i.e. waiting for a response.
     pub fn is_pending_outbound(&self, peer: &PeerId, request_id: &RequestId) -> bool {
-        self.connected.get(peer)
+        // check if request id is exists in already established connections
+        let est_conn = self.connected.get(peer)
             .map(|cs| cs.iter().any(|c| c.pending_inbound_responses.contains(request_id)))
-            .unwrap_or(false)
+            .unwrap_or(false);
+        // check if peer exists in pending connections
+        let pen_conn = self.pending_outbound_requests.contains_key(peer);
```

Comment:
Say you have two requests R1 and R2 all targeting peer P1. Say R1 is done and a result is returned to the upper layer. Say R2 is still in `self.pending_outbound_requests`.

If you call `is_pending_outbound(P1, R1)` you will get `true` even though R1 is already finished. You need to check the `request_id` in the `RequestProtocol` in `pending_outbound_request`.

Reply:
Should we include this scenario in the test program, too? I think, we should.

Reply:
I added a few more assert statements in ping.rs to cover this scenario. Thank you for bringing up this. :)

Reply:
I'm trying to replicate this scenario. For some reason, I'm not able to do so.

I added a few print statements, to log stuff to the console, in the `is_pending_outbound` method.
```rust
let est_conn = self.connected.get(peer)
    .map(|cs| cs.iter().any(|c| c.pending_inbound_responses.contains(request_id)))
    .unwrap_or(false);

let pen_conn = self.pending_outbound_requests.contains_key(peer);

println!("request_id (from is_pending_outbound): {}", *request_id);
println!("est_conn && pen_conn: {} || {}", est_conn, pen_conn);
// print all the request ids in pending_outbound_requests for a given peer id
for value in self.pending_outbound_requests.get(peer) {
    for item in value {
        println!("request_id (from pending_outbound_requests): {}", item.request_id);
    }
}

est_conn || pen_conn
```

And, then I called the `is_pending_outbound` method in 'ping_protocol` method in `ping.rs` as follows:
```rust
let mut req_id = swarm2.send_request(&peer1_id, ping.clone());
let mut req_id_org = req_id.clone();
assert!(swarm2.is_pending_outbound(&peer1_id, &req_id));

let peer2 = async move {
    let mut count = 0;
    let addr = rx.next().await.unwrap();
    swarm2.add_address(&peer1_id, addr.clone());
    let mut req_id = swarm2.send_request(&peer1_id, ping.clone());
    let mut req_id_org = req_id.clone();
    assert!(swarm2.is_pending_outbound(&peer1_id, &req_id));

    //let mut req_id2 = swarm2.send_request(&peer1_id, ping.clone());
    loop {
        match swarm2.next().await {
            RequestResponseEvent::Message {
                peer,
                message: RequestResponseMessage::Response { request_id, response }
            } => {
                count += 1;
                assert_eq!(&response, &expected_pong);
                assert_eq!(&peer, &peer1_id);
                assert_eq!(req_id, request_id);

                // check if is_pending_outbound returns false for a finished request when we have already issued a new one
                if req_id != req_id_org {
                    assert!(swarm2.is_pending_outbound(&peer1_id, &req_id_org) == false);
                }
                
                req_id_org = req_id.clone();
                if count >= num_pings {
                   break 
                } else {
                    req_id = swarm2.send_request(&peer1_id, ping.clone());
                    assert!(swarm2.is_pending_outbound(&peer1_id, &req_id));
                }
            },
            e => panic!("Peer2: Unexpected event: {:?}", e)
        }
    }
```

On running this code, I get the following output:
```
request_id (from is_pending_outbound): 2
est_conn && pen_conn: true || false
request_id (from is_pending_outbound): 1
est_conn && pen_conn: false || false
request_id (from is_pending_outbound): 3
est_conn && pen_conn: true || false
request_id (from is_pending_outbound): 2
est_conn && pen_conn: false || false
request_id (from is_pending_outbound): 4
est_conn && pen_conn: true || false
request_id (from is_pending_outbound): 3
```

So, in every iteration of the loop, I'm checking if the previous request is outbound and it **always returns a false statement**. So, I guess, the edge case you mentioned doesn't occur? I highly suspect that I am doing something wrong.